processing/src/main/java/org/apache/druid/frame/write/cast/TypeCastSelectors.java

```diff
@@ -0,0 +1,149 @@
+        desiredType
+    );
+  }
+
+  public static ColumnValueSelector<?> wrapColumnValueSelectorIfNeeded(
+      final ColumnValueSelector<?> selector,
+      @Nullable final ColumnCapabilities selectorCapabilities,
+      @Nullable final RowIdSupplier rowIdSupplier,
+      final ColumnType desiredType
+  )
+  {
+    if (desiredType.is(ValueType.STRING)) {
+      throw DruidException.defensive("Type[%s] should be read using a DimensionSelector", desiredType);
+    } else if (desiredType.isNumeric()
+               && (selectorCapabilities == null || !selectorCapabilities.isNumeric())) {
+      // When capabilities are unknown, or known to be non-numeric, fall back to getObject() and explicit typecasting.
+      // This avoids using primitive numeric accessors (getLong / getDouble / getFloat / isNull) on a selector that
+      // may not support them.
+      return new ObjectToNumberColumnValueSelector(selector, desiredType, rowIdSupplier);
+    } else if (desiredType.isArray()) {
+      // Always wrap if desiredType is an array. Even if the underlying selector claims to offer the same type as
+      // desiredType, it may fail to respect the BaseObjectColumnValueSelector contract. For example, it may return
+      // List rather than Object[]. (RowBasedColumnSelectorFactory can do this if used incorrectly, i.e., if the
+      // ColumnInspector declares type ARRAY<X> for a column, but the RowAdapter does not provide Object[].)
+      return new ObjectToArrayColumnValueSelector(selector, desiredType, rowIdSupplier);
+    } else {
+      // OK to return the original selector.
+      return selector;
+    }
+  }
+
+  /**
+   * Coerce an object to an object compatible with what {@link BaseObjectColumnValueSelector#getObject()} for a column
+   * of the provided desiredType. Never throws an exception. If coercion fails, replaces the object that failed to
+   * coerce with null.
+   *
+   * @param obj         object
+   * @param desiredType desired type
+   */
+  @Nullable
+  public static Object bestEffortCoerce(
+      @Nullable final Object obj,
+      @Nullable final TypeSignature<ValueType> desiredType
+  )
+  {
+    if (obj == null || desiredType == null) {
+      return obj;
+    }
+
+    ValueType type = desiredType.getType();
+
+    if (type == ValueType.STRING) {
+      return ExprEval.bestEffortOf(obj).asString();
+    } else if (type == ValueType.LONG) {
+      final ExprEval<?> n = ExprEval.bestEffortOf(obj);
+      return n.isNumericNull() ? null : n.asLong();
+    } else if (type == ValueType.DOUBLE) {
+      final ExprEval<?> n = ExprEval.bestEffortOf(obj);
+      return n.isNumericNull() ? null : n.asDouble();
+    } else if (type == ValueType.FLOAT) {
+      final ExprEval<?> n = ExprEval.bestEffortOf(obj);
+      return n.isNumericNull() ? null : (float) n.asDouble();
+    } else if (type == ValueType.ARRAY) {
+      final TypeSignature<ValueType> elementType = desiredType.getElementType();
+
+      if (obj instanceof List) {
```

Comment:
you could use `ExprEval.bestEffortArray` here i suppose, though `ExprEval.ofType` if desiredType is an array, as well as `ExprEval.bestEffortOf` should have pretty similar logic.